src/main/java/com/scaleunlimited/flinkcrawler/sources/SeedUrlSource.java

```diff
@@ -130,25 +129,53 @@ public void cancel() {
 	public void run(SourceContext<RawUrl> context) throws Exception {
 		_keepRunning = true;
 		
+        BufferedReader s3FileReader = null;
+        if (useS3File()) {
+            s3FileReader = new BufferedReader(new InputStreamReader(_s3FileStream));
+        }
+        
 		while (_keepRunning) {
+		    
+	        // TODO Handle skipping empty lines/comments after we've decided if
+	        // the line is for the appropriate partition, so that we could use
+	        // _seedUrlIndex to re-sync if we get terminated/checkpointed/the
+	        // stream is closed & we have to reopen it, etc.
+		    // 
+		    // Hmmm. That would make _seedUrlIndex an input line index rather
+		    // than a URL index, no?  Assuming the file contents don't change,
+		    // it seems like we could either save a separate input line index
+		    // (i.e., to save re-parsing each input line for the re-sync),
+		    // or we just let it spin through the file re-parsing each line
+		    // until we've seen _seedUrlIndex real URLs.
+		    //
 			if (useS3File()) {
-				// TODO read the next line from the file. Break if we're at the end.
-				// Handle skipping empty lines/comments after we've decided if the
-				// line is for the appropriate partition, so that we could use the
-				// _seedUrlIndex to re-sync if we get terminated/checkpointed/the
-				// stream is closed & we have to reopen it, etc.
+			    String sourceLine = s3FileReader.readLine();
+			    if (sourceLine == null) {
+			        break;
+			    }
+			    RawUrl url = parseSourceLine(sourceLine);
+			    if (url != null) {
+	                collectUrl(url, context);
```

Comment:
If we have parallelism in the source, then won't this wind up reading the file N times, and emitting the same seed URL N times?